File: 백준/Gold/12738. 가장 긴 증가하는 부분 수열 3/가장 긴 증가하는 부분 수열 3.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(bf.readLine());
        int[] numbers = Arrays.stream(bf.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        List<Integer> lis = new ArrayList<>();
        lis.add(numbers[0]);

        for(int i=1; i<N; i++){
            if(lis.get(lis.size()-1) < numbers[i]){
                lis.add(numbers[i]);
            } else {
                int idx = binarySearch(0, lis.size()-1, numbers[i], lis);
                lis.set(idx, numbers[i]);
            }
        }

        System.out.println(lis.size());
    }

    public static int binarySearch(int left, int right, int target, List<Integer> list){
        while(left < right){
            int mid = (left + right) / 2;
            if(list.get(mid) < target){
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return right;
    }
}

File: 백준/Gold/12738. 가장 긴 증가하는 부분 수열 3/가장 긴 증가하는 부분 수열 3.cc
#include <iostream>
#include <vector>

using namespace std;

int bs(vector<int>& list, int num){

    int left=0, right=list.size(), mid;
    while (left < right){
        mid = (left+right)/2;

        if(list[mid]<num)
            left = mid+1;
        else{
            right = mid;
        }
    }
    return right;
}

int main() {
    int n;
    cin>> n;
    vector<int> nums(n);

    for(int i=0;i<n;i++){
        cin>> nums[i];
    }

    vector<int> list;
    list.push_back(nums[0]);

    for(int i=1;i<n;i++){
        if(list.back() < nums[i]){
            list.push_back(nums[i]);
        }
        else{
            int idx = bs(list, nums[i]);
            list[idx] = nums[i];
        }
    }

    cout<<list.size();

    return 0;
}



